subprojects/vinumc/vec.h

```diff
@@ -27,12 +27,29 @@
 		(arr)->len++;                                                                      \
 	} while (0)
 
-#define VEC_RESERVE(arr, size)                                                                     \
+#define VEC_RESERVE_EXACT(arr, size)                                                               \
 	do {                                                                                       \
 		(arr)->capacity = (size);                                                          \
 		(arr)->base = realloc((arr)->base, (arr)->capacity * sizeof(*(arr)->base));        \
 	} while (0)
 
+#define VEC_RESERVE(arr, size)                                                                     \
+	do {                                                                                       \
+		size_t needed = (arr)->len + size;                                                 \
+		if (needed <= (arr)->capacity)                                                     \
+			break;                                                                     \
+		/* rounds needed to the nearest power of 2 >= its current value */                 \
+		needed--;                                                                          \
+		needed |= needed >> 1;                                                             \
+		needed |= needed >> 2;                                                             \
+		needed |= needed >> 4;                                                             \
+		needed |= needed >> 8;                                                             \
+		needed |= needed >> 16;                                                            \
+		needed |= needed >> 32;                                                            \
```

Comment:
this is not portable, right? What about turning this into a loop?


```suggestion

		for (int i = 1; i < sizeof(size_t); i <<= 1) \
			needed |= needed >> i; \                                                       
```